[java] DataflowAnomalyAnalysis Rule in right order

YoonhoChoi96 opened the pull request:
<!--
Please, prefix the PR title with the language it applies to within brackets, such as *[java]* or *[apex]*. If not specific to a language, you can use *[core]*
-->

Before submitting a PR, please check that:
 - [v] The PR is submitted against `master`. The PMD team will merge back to support branches as needed.
 - [v] `./mvnw clean verify` passes. This will [build](https://github.com/pmd/pmd/blob/master/BUILDING.md) and test PMD, execute PMD and checkstyle rules. [Check this for more info](https://github.com/pmd/pmd/blob/master/CONTRIBUTING.md#code-style)

Fixes   #1749, I checked this issue and I fixed as follows. Please, review this Pull Request

- Added a logic that could make DataflowAnomlayAnalysis Rule check codes in right order.

- When assignment is occurred between same varaibles, it read left to right without caring of the precedence. In my opinion, it would make a many DD False Positive. Therefore, I added the logic which could check and swap an order of elements to maintain rigth order.

- After this modification, I had a time to check that it works in various code skeletons and open source project. As a result of testing code skeletons, current version of pmd make 11 alarms and 9 of these alarm was false positive. However, revised version of pmd make 11 alarms and 1 of these alarma was false positive(You could find test code skeletons below). And, as the result of testing open source project(MaterialDesign), it reduced 17 False Positives(Current : 40, Revised 23).

- Although it is just small part, I hope that it may help to reduce number of false positives in DataflowAnomalyAnalysis Rule. 


- Test Code Skeletons **(False Positive as Bold)**

Case 1.
```
class Test{
   public static void main(String[] args){
       int a = 0 ;
       int b = 0 ;
       a = a + b ;
       b = a + b ;
       }
}
```
> Org. : **DD(a)** / Rev. : DU(b)

Case 2.
```
class Test{
   public static void main(String[] args){
       int a = 0 ;
      
       for(int i = 0 ; i <= 10; i ++){
           a = a+3;
       }
   }
}
```
> Org. : **DD(a)**, DU(a) / Rev. : DU(a)

Case 3.
```
class Test{
   public static void main(String[] args){
      
       int[] b = new int[10];

       for(int a : b){
           a = a+3;
       }
   }
}
```
> Org. : No Alarm / Rev. : **UR(a)**, DU(a)

Case 4.
```
class Test{
   public static void main(String[] args){
       int a = 0 ;
       int i = 0 ;
       while(i < 30){
           a = a+3;
           i += 3;
       }
   }
}

```
> Org. : **DD(a)**, DU(a) / Rev. : DU(a)

Case 5.
```
class Test{
   public static void main(String[] args){
       int a = 0 ;
       int i = 0 ;
       do{
           a = a+3;
           i += 3;
       }while(i < 30);
   }
}
```
> Org. : **DD(a)** / Rev. : DU(a)

Case 6.
```
class Test{
   public static void main(String[] args){
       int a = 0 ;
       int i = 0 ;
       switch(i){
           case 1 : a = a+1;
           break;
           case 2 : a = a+2;
           break;
           case 3 : a = a+3;
           break;
           default : a = a + 0;
       }
   }
}
```
> Org. : **DD(a)** / Rev. : DU(a)

Case 7.
```
class Test{
   public static void main(String[] args){
       int a = 0 ;
       int i = 0 ;
       switch(i){
           case 1 : a = a+1;
           case 2 : a = a+2;
           case 3 : a = a+3;
           default : a = a + 0;
       }
   }
}
```
> Org. : **DD(a)** / Rev. : DU(a)

Case 8.
```
class Test{
   public static void main(String[] args){
       Test2 test = new Test2() ;
       test.a = test.b;
   }
}
```
> Org. : No Alarm / Rev. : No Alarm

Case 9.
```
class Test{
   public static void main(String[] args){
       int t1 = 0 ;
       Test2 test = new Test2() ;
       t1 = test.simpleTest(t1) ;
   }
}
```
> Org. : **DD(t1)** / Rev. : DU(t1)

Case 10.
```
class Test{
   public static void main(String[] args){
       Test2 test = new Test2() ;
       test.a = test.simpleTest(test.a) ;
   }
}
```
> Org. : No Alarm / Rev. : No Alarm

Case 11.
```
class Test{
   public static void main(String[] args){
       Test2 test = new Test2() ;
       test.a = 3 ;
       test.a = 5 ;
   }
}
```
> Org. : No Alarm / Rev. : No Alarm

Case 12.
```
class Test{
   public static void main(String[] args){
       int t1 = 0 ;
       int t2 = 0 ;
       Test2 test = new Test2();
       if((t1 = test.simpleTest(t1)) == t2);
   }
}
```
> Org. : **DD(t1)** / Rev. : DU(t1)

Case 13.
```
class Test{
   public static void main(String[] args){
       int t1 = 0 ;

       Test2 test = new Test2();
       if( (t1 = test.simpleTest(t1)) == t1);
   }
}
```
> Org. : **DD(t1)** / No Alarm

Case 14.
```
class Test{
   public static void main(String[] args){
       int t1 = 0;
       int t2 = 0;
       Test2 test = new Test2() ;
       if( t1 == (t1 = test.simpleTest(t1))) ;
   }
}
}
```
> Org. : No Alarm / Rev. : DU(t1)

pmd-test commented:
<!--
  0 Errors
  0 Warnings
  1 Message: This changeset introduces 184 ...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">This changeset introduces 184 new violations, 0 new errors and 0 new configuration errors,<br />
removes 550 violations, 0 errors and 0 configuration errors.<br />
<a href="https://chunk.io/pmd/382bebb54e344ec69cbef11f3b1dbc47/diff/index.html">Full report</a></td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

adangel commented on pmd-java/src/main/java/net/sourceforge/pmd/lang/java/dfa/VariableAccessVisitor.java:
```diff
@@ -76,8 +80,58 @@ private void computeNow(Node node) {
+                    for (Node temp = name.getLocation(); !(temp.jjtGetParent() instanceof ASTMethodDeclaration) 
+                        && !(temp.jjtGetParent() instanceof ASTConstructorDeclaration); temp = temp.jjtGetParent()) {
+                        //traverse children nodes to check it has assignment operator as a child 
+                        for (int i = 0; i < temp.jjtGetNumChildren(); i++) {
```
You can simplify this loop like this:

```java
List<Node> assignmentOperators = temp.findChildrenOfType(ASTAssignmentOperator.class);
if (!assignmentOperators.isEmpty()) {
 ...
}
```

YoonhoChoi96 commented:
Thank you for your kind reply.

I added test cases to the file. I only added about false positive cases.
It means that I didn't add about 'no alarm' cases.
Because it was for catch False Positives, 
and I couldn't sure about expected result of 'no alarm' cases.
And Also, I modified nested loop as your suggestion.
When I try to run it, It occurs error like

`incompatible types: java.util.List<net.sourceforge.pmd.lang.java.ast.ASTAssignmentOperator> cannot be converted to java.util.List<net.sourceforge.pmd.lang.ast.Node>`

So, I changed little bit. `Node` -> `ASTAssignmentOperator` like below.

`List<ASTAssignmnetOperator> assigmentOperators = temp.findChilderenOfType(ASTAssignmentOperator.class`

Please, review about it.

And I think you fully understood about the problem what I want to express.

Thank you for rephrasing it and your great suggestion.

YoonhoChoi96 commented:
Previous commits had wrong `expected problems` number, 
so, I fixed it to pass build process.
And the case which include for-each statement is commented out.
becauses,  as I know, for-each statement is not fully applied in this tool.
So, I just comments out the case.

Thank you.

adangel commented on pmd-java/src/main/java/net/sourceforge/pmd/lang/java/dfa/VariableAccessVisitor.java:
```diff
@@ -76,8 +80,58 @@ private void computeNow(Node node) {
                 }
                 undefinitions.add(new VariableAccess(VariableAccess.UNDEFINITION, vnd.getImage()));
 
+                //make arraylist which could store all name occurrences
```
We don't need so many lists... We have essentially two stages:
1. Map the name occurrences to their assignment expressions, if there are any -> `List<SimpleEntry<..,..>>`
2. Sort the name occurrences -> same List

Last step is to add the variable access in order -> we simply traverse the same list, no need to create a new one.

I've come up with the following code which only create one temporary list (occurrencesWithAssignmentExp):

``` java
                // Map the name occurrences to their assignment expressions, if any
                List<SimpleEntry<Node, NameOccurrence>> occurrencesWithAssignmentExp = new ArrayList<>();
                for (NameOccurrence occurrence : entry.getValue()) {
                    // find the nearest assignment, if any
                    Node potentialAssignment = occurrence.getLocation().getFirstParentOfAnyType(ASTStatementExpression.class,
                            ASTExpression.class);
                    while (potentialAssignment != null
                            && (potentialAssignment.jjtGetNumChildren() < 2
                                    || !(potentialAssignment.jjtGetChild(1) instanceof ASTAssignmentOperator))) {
                        potentialAssignment = potentialAssignment.getFirstParentOfAnyType(ASTStatementExpression.class,
                                ASTExpression.class);
                    }
                    // at this point, potentialAssignment is either a assignment or null
                    occurrencesWithAssignmentExp.add(new SimpleEntry<>(potentialAssignment, occurrence));
                }

                // The name occurrences are in source code order, that means, the left hand side of the
                // assignment is first. But this is not the order in which the data flows: first the
                // right hand side is evaluated before the left hand side is assigned.
                // Therefore move the name occurrences backwards if they belong to the same assignment expression.
                for (int i = 0; i < occurrencesWithAssignmentExp.size() - 1; i++) {
                    SimpleEntry<Node, NameOccurrence> oc = occurrencesWithAssignmentExp.get(i);
                    SimpleEntry<Node, NameOccurrence> nextOc = occurrencesWithAssignmentExp.get(i + 1);

                    if (oc.getKey() != null && oc.getKey().equals(nextOc.getKey())) {
                        Collections.swap(occurrencesWithAssignmentExp, i, i + 1);
                    }
                }

                for (SimpleEntry<Node, NameOccurrence> oc : occurrencesWithAssignmentExp) {
                    addAccess((JavaNameOccurrence) oc.getValue(), inode);
                }
```

I used `getFirstParentOfAnyType` to directly go to potential assignment expressions. Assignment expressions are either Expression nodes or StatementExpression node, which have the assignment operator as a second child.

YoonhoChoi96 replied:
This solution extremly simplify the codes!
I've learned lots of things from this chance:)
I modified the codes according to your instruction,
and I think it looks far much better.
Thanks a lot.

YoonhoChoi96 commented on pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/rule/errorprone/xml/DataflowAnomalyAnalysis.xml:
```diff
@@ -132,6 +132,186 @@ public class AssertTest {
+        ]]></code>
+    </test-code>
+
+    <test-code>
```
Oh, I didn't consider that. I checked it and I added. Thanks for kind instruction!

YoonhoChoi96 commented:
Thanks a lot!
